### PR TITLE
Avoid ranging over recursive dependencies

### LINKED ISSUE #46: Infinite loop on circular dependency
**Describe the bug**
In case of circular dependency withing the target of zconfig.Configure, the function is caught in an infinite loop.
This has been detected using `gorm.io/gorm.DB v1.22.4`

**To Reproduce**
```
package main

import (
        "fmt"

        "github.com/synthesio/zconfig"
        "gorm.io/gorm"
)

type Service struct {
        *gorm.DB
}

func main() {
        fmt.Println("start")
        zconfig.Configure(new(Service))
        fmt.Println("done")
}
```

**Expected behavior**
We expect zconfig to properly walk through the service, whether it embeds a circular dependency or not.

### PR DESCRIPTION
**Pull request by elwinar**
Fixes #46.

### CODE REVIEW

File: processor.go [hunk 1]
```diff
@@ -114,12 +114,33 @@ func walk(v reflect.Value, s reflect.StructField, p *Field) (field *Field, err e
...
+		for ancestor := field; ancestor != nil; ancestor = ancestor.Parent {
+			ancestorType := ancestor.Value.Type()
+			for ancestorType.Kind() == reflect.Ptr {
+				ancestorType = ancestorType.Elem()
+			}
+
+			fieldType := structField.Type
+			for fieldType.Kind() == reflect.Ptr {
+				fieldType = fieldType.Elem()
+			}
```
Comment (dsiragusa) [inline processor.go:137]: This could be done outside the ancestor loop

File: processor_test.go [hunk 2]
```diff
@@ -20,9 +21,28 @@ type SimpleDependency struct {
...
+type RecursiveDependency struct {
+	Service
+}
+
 func TestWalk(t *testing.T) {
 	var v = reflect.ValueOf(new(Service))
-	root, err := walk(v, reflect.StructField{}, nil)
+
+	// The check itself is done in non-blocking fashion so we can fail in
+	// case of dependency loop, which can happen when embeding recursive
```
Comment (dsiragusa) [inline processor_test.go:32]: Spelling issue for embedding

File: processor.go (hunk 1 above)
Comment (elwinar) [reply processor.go]: Fixed.

File: processor_test.go (hunk 2 above)
Comment (elwinar) [reply processor_test.go]: Fixed.

**Review by dsiragusa — APPROVED** [review]

**Pull request merged**